lib/Dialect/Torch/Transforms/MaximizeValueSemantics.cpp

```diff
@@ -37,136 +37,169 @@ class AbstractlyInterpretCopyToNonValueTensorOpUsersWithinABlock
+          return rewriter.notifyMatchFailure(
+              copyToNonValueTensor, "overwritten tensor is not a valid alias");
+        }
+
+        // To simplify the analysis, we only support the case where the
+        // only aliases used after an overwrite are the aliases generated
+        // after plus the alias being overwritten.
+        availableAliases.clear();
+        availableAliases.insert(overwritten);
+        result.overwriteTensorContentsOps.push_back(overwrite);
       } else {
-        llvm_unreachable("only those ops supported!");
+        return rewriter.notifyMatchFailure(
+            copyToNonValueTensor,
+            "unsupported op encountered during abstract analysis");
       }
     }
-    rewriter.eraseOp(copy);
-    return success();
+    return result;
+  }
+
+  // Rewrite slice composed of the interpreted ops so that the slice uses
+  // value semantics everywhere.
+  static void rewriteSlice(const InterpretedOps &ops,
+                           PatternRewriter &rewriter) {
+    // The rewriting for the overwrite op involves replacing all uses of its
+    // non-value tensor operand with its value tensor operand. Since the
+    // rewriting of other ops can potentially change the non-value tensor
+    // operand to a value tensor, this rewriting MUST happen first to avoid
+    // wrongly replacing operands that were previously not a view of the
+    // overwritten tensor.
+    for (OverwriteTensorContentsOp overwrite :
+         llvm::reverse(ops.overwriteTensorContentsOps)) {
+      Value overwritten = overwrite.overwritten();
+      assert(overwritten.getType().dyn_cast<NonValueTensorType>() &&
+             "the analysis assumes that overwritten remains a nonValueTensor "
+             "throughout the rewriting");
+      overwritten.replaceUsesWithIf(
+          overwrite.value(), [&](const OpOperand &operand) {
+            return !operand.getOwner()->isBeforeInBlock(overwrite);
+          });
+      rewriter.eraseOp(overwrite);
+    }
+
+    for (Operation *copyLikeOp : ops.copyLikeOps)
+      rewriter.replaceOp(copyLikeOp, copyLikeOp->getOperand(0));
+
+    // Replace return type of view-like ops with value-semantics type variant.
+    for (Operation *viewLikeOp : ops.viewLikeOps) {
+      rewriter.updateRootInPlace(viewLikeOp, [&] {
+        Value result = viewLikeOp->getResult(0);
+        auto resultType = result.getType().dyn_cast<NonValueTensorType>();
+        assert(resultType && "all view-like ops considered must have result of "
+                             "type `NonValueTensorType` before rewriting");
+        result.setType(resultType.getWithValueSemantics());
+      });
+    }
   }
-};
-} // namespace
 
-namespace {
-// Calculate a forward slice starting from a CopyToNonValueTensorOp
-// and ending at CopyToValueTensorOp's. If all intervening ops
-// are just view-like operations (i.e. no mutation), then we can trivially
-// convert them all to value semantics.
-class RewriteViewLikeSubgraph
```

Comment:
The `RewriteViewLikeSubgraph` still has value in cases where the use-def tree spanning multiple blocks with branches but with no writes. We can keep it until a more sophisticated approach is implemented.

Reply:
I think that the improved version that Ramiro is working on should handle this case too. Looking forward to having these two fully unified!

Reply:
I've added back the `RewriteViewLikeSubgraph` in the meantime while I work on improving the abstract interpretation pas